File: Algorithms_Part_1/Algorithm1Sample/Ex35LinearProbingHashST.java
/* *****************************************************************************
 *  Name:
 *  Date:
 *  Description:
 **************************************************************************** */

import edu.princeton.cs.algs4.Queue;
import edu.princeton.cs.algs4.StdOut;

public class Ex35LinearProbingHashST<Key, Value> {

    private static final int M = 1000;
    private Key[] keys = (Key[]) new Object[M];
    private Value[] values = (Value[]) new Object[M];
    private int sz = 0;

    public void put(Key key, Value value) {
        int i = hash(key), tombStone = -1;
        while (keys[i] != null && !keys[i].equals(key)) {
            if (tombStone == -1 && values[i] == null) tombStone = i;
            i = nextIndex(i);
        }
        if (keys[i] == null) {
            if (tombStone != -1) i = tombStone;
            keys[i] = key;
            values[i] = value;
            sz++;
        }
        else values[i] = value;
    }

    public Value get(Key key) {
        int i = hash(key);
        while (keys[i] != null && !keys[i].equals(key)) i = nextIndex(i);
        if (keys[i] == null) return null;
        return values[i];
    }

    public void delete(Key key) {
        int i = hash(key);
        while (keys[i] != null && !keys[i].equals(key)) i = nextIndex(i);
        if (keys[i] == null) return;
        values[i] = null;
        sz--;
    }

    public boolean contains(Key key) {
        return get(key) != null;
    }

    public boolean isEmpty() {
        return size() == 0;
    }

    public int size() {
        return sz;
    }

    public Iterable<Key> keys() {
        Queue<Key> queue = new Queue<>();
        for (int i = 0; i < M; i++) {
            if (keys[i] != null && values[i] != null) queue.enqueue(keys[i]);
        }
        return queue;
    }

    private int hash(Key key) {
        return (key.hashCode() & 0x7fffffff) % M;
    }

    private int nextIndex(int i) {
        return (i + 1) % M;
    }

    public static <Key extends Comparable<Key>> void test(Key[] keys, Key lo, Key hi, Key floor1,
                                                          Key floor2) {

        Ex35LinearProbingHashST<Key, Integer> bstTestCase
                = new Ex35LinearProbingHashST<>();
        int index = 1;
        for (Key i : keys) {
            StdOut.print("(" + i + "," + index + ") ");
            bstTestCase.put(i, index++);
        }
        StdOut.println();
        StdOut.println("Size: " + bstTestCase.size());
        StdOut.print("Keys : ");
        for (Key i1 : bstTestCase.keys()) {
            StdOut.print(i1 + " ");
        }
        StdOut.println();
        StdOut.println("get " + lo + ":" + bstTestCase.get(lo));
        StdOut.println("get " + hi + ":" + bstTestCase.get(hi));
        StdOut.println("get " + floor1 + ":" + bstTestCase.get(floor1));
        StdOut.println("get " + floor2 + ":" + bstTestCase.get(floor2));
        StdOut.print("After Deleting " + lo + ": ");
        bstTestCase.delete(lo);
        for (Key i1 : bstTestCase.keys()) {
            StdOut.print(i1 + " ");
        }
        StdOut.println();
        StdOut.print("After Deleting " + hi + ": ");
        bstTestCase.delete(hi);
        for (Key i1 : bstTestCase.keys()) {
            StdOut.print(i1 + " ");
        }
        StdOut.println();
    }

    public static void main(String[] args) {
        StdOut.println("Test Case 1 :");
        Integer[] testCase = { 54, 5, -9, 53, 1, 0, 15, -9 };
        test(testCase, 0, 20, -9, 15);

        StdOut.println();
        StdOut.println("Test Case 2 :");
        Character[] testCase2 = { 'S', 'O', 'R', 'T', 'E', 'X', 'A', 'M', 'P', 'L', 'E' };
        test(testCase2, 'B', 'P', 'E', 'P');
    }
}

File: PythonLearnig/Algorithms_Part_1/Ex35LinearProbingHashST.py
class LinearProbingHashST:

    M = 100

    def __init__(self):
        self.all_keys = [None] * self.M
        self.values = [None] * self.M
        self.sz = 0

    def put(self, key, value):

        i = self.hash(key)
        while True:
            if self.all_keys[i] is None:
                self.all_keys[i] = key
                self.values[i] = value
                self.sz += 1
                break
            elif self.all_keys[i] is key:
                self.values[i] = value
                break
            i = self.next_index(i)

    def get(self, key):

        i = self.hash(key)
        while self.all_keys[i] is not None:
            if self.all_keys[i] is key:
                return self.values[i]
            i = self.next_index(i)
        return None

    def contains(self, key):
        return self.get(key) is not None

    def size(self):
        return self.sz

    def is_empty(self):
        return self.size() == 0

    def keys(self):
        for key in self.all_keys:
            if key is not None:
                yield key

    def delete(self, key):

        key_hash = self.hash(key)

        if self.all_keys[key_hash] is None:
            return None

        i = key_hash
        while self.all_keys[i] is not None and self.all_keys[i] is not key:
            i = self.next_index(i)
            if i == key_hash:
                return None

        value = self.values[i]
        next_i = self.next_index(i)
        self.all_keys[i] = None
        self.values[i] = None
        self.sz -= 1

        while self.all_keys[next_i] is not None and next_i != key_hash:

            if self.hash(self.all_keys[next_i]) == key_hash:
                self.all_keys[i] = self.all_keys[next_i]
                self.values[i] = self.values[next_i]
                self.all_keys[next_i] = None
                self.values[next_i] = None
                i = next_i

            next_i = self.next_index(i)

        return key, value

    def hash(self, key):
        return (hash(key) * 0x7fffffff) % self.M

    def next_index(self, i):
        return (i + 1) % self.M

    def __str__(self):

        result = '\n'
        for i in range(self.M):
            if self.all_keys[i] is not None:
                result += f'Key @ ({i}) : {self.all_keys[i]} Value: {self.values[i]}\n'
        return result


def test_linear_probing_hash_st():

    def test(keys, test_keys):
        import random

        st = LinearProbingHashST()
        index = 1
        for key in keys:
            st.put(key, index)
            print(f'Insert {key} = {index} : {list(st.keys())} size: {st.size()}')
            index += 1

        print('='*10, st, '='*10)
        print(f'Key,Value: {[(key, st.get(key)) for key in st.keys()]}')
        print(f"Size: {st.size()}")
        print("Keys:", list(st.keys()))
        for key in test_keys:
            print(f'Contains {key}? {st.contains(key)} Get {key}: {st.get(key)}')
            st.put(key, random.randint(500, 600))
        for key in test_keys:
            print(f"After deleting key {key}: {st.delete(key)} keys: {list(st.keys())}")

        print('='*10, st, '='*10)
        print('Size:', st.size())
        print()
        print()

    print("Test case 1:")
    test_case = [54, 5, -9, 53, 1, 0, 15]
    test(test_case, [0, 20, 10, 15])

    print("Test case 2:")
    test_case = ['S', 'O', 'R', 'T', 'E', 'X', 'A', 'M', 'P', 'L', 'E']
    test(test_case, ['B', 'P', 'G', 'P'])

    print(' END '.center(100, '-'))


if __name__ == '__main__':
    test_linear_probing_hash_st()
